pkg/util/kubernetes/autoscalers/datadogclient.go

```diff
@@ -0,0 +1,234 @@
+	endpoint := os.Getenv("DATADOG_HOST")
+	if config.Datadog.GetString(metricsEndpointConfig) != "" || endpoint == "" {
+		endpoint = config.GetMainEndpoint(metricsEndpointPrefix, metricsEndpointConfig)
+	}
+
+	if appKey == "" || apiKey == "" {
+		return nil, errors.New("missing the api/app key pair to query Datadog")
+	}
+
+	log.Infof("Initialized the Datadog Client for HPA with endpoint %q", endpoint)
+
+	client := datadog.NewClient(apiKey, appKey)
+	client.HttpClient.Transport = httputils.CreateHTTPTransport()
+	client.RetryTimeout = 3 * time.Second
+	client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+	client.SetBaseUrl(endpoint)
+
+	return client, nil
+}
+
+type datadogIndividualClient struct {
+	client             *datadog.Client
+	lastQuerySucceeded bool
+	lastFailure        time.Time
+	lastSuccess        time.Time
+	retryInterval      time.Duration
+}
+
+const (
+	minRetryInterval = 30 * time.Second
+	maxRetryInterval = 30 * time.Minute
+)
+
+// DatadogFallbackClient represents a datadog client able to query metrics to a second Datadog endpoint if the first one fails
+type datadogFallbackClient struct {
+	clients        []*datadogIndividualClient
+	lastUsedClient int
+}
+
+// NewDatadogFallbackClient generates a new client able to query metrics to a second Datadog endpoint if the first one fails
+func newDatadogFallbackClient(endpoints []config.Endpoint) (*datadogFallbackClient, error) {
+	if len(endpoints) == 0 {
+		return nil, log.Errorf("external_metrics_provider.endpoints must be non-empty")
+	}
+
+	defaultClient, err := newDatadogSingleClient()
+	if err != nil {
+		return nil, err
+	}
+
+	ddFallbackClient := &datadogFallbackClient{
+		clients: []*datadogIndividualClient{
+			{
+				client:             defaultClient,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			},
+		},
+	}
+	for _, endpoint := range endpoints {
+		client := datadog.NewClient(endpoint.APIKey, endpoint.APPKey)
+		client.HttpClient.Transport = httputils.CreateHTTPTransport()
+		client.RetryTimeout = 3 * time.Second
+		client.ExtraHeader["User-Agent"] = "Datadog-Cluster-Agent"
+		client.SetBaseUrl(endpoint.URL)
+		ddFallbackClient.clients = append(
+			ddFallbackClient.clients,
+			&datadogIndividualClient{
+				client:             client,
+				lastQuerySucceeded: true,
+				retryInterval:      minRetryInterval,
+			})
+	}
+
+	return ddFallbackClient, nil
+}
+
+func (ic *datadogIndividualClient) queryMetrics(from, to int64, query string) ([]datadog.Series, error) {
+	// TODO: check for err 500 (server side) instead of any kind of error because switching to DR doesn’t make sense if that’s the query which is bad.
+	if series, err := ic.client.QueryMetrics(from, to, query); err == nil {
+		ic.lastQuerySucceeded = true
+		ic.lastSuccess = time.Now()
+		ic.retryInterval /= 2
+		if ic.retryInterval < minRetryInterval {
+			ic.retryInterval = minRetryInterval
+		}
+		return series, err
+	} else { // revive:disable-line:indent-error-flow
```

Comment:
Instead of disabling lint, would it make sense to move duplicated return statement after the `if` body?